packages/rpc-methods/src/utils.ts

```diff
@@ -4,31 +4,31 @@ export function isPlainObject(
   return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
 }
 
-type SelectedHooks<
-  Hooks extends Record<keyof HookNames, unknown>,
-  HookNames extends Record<string, boolean>,
-> = Pick<Hooks, keyof HookNames & keyof Hooks>;
-
-export function selectHooks<Hooks extends Record<string, unknown>>(
-  hooks: Hooks,
-  hookNames?: void,
-): void;
-
+/**
+ * Returns the subset of the specified `hooks` that are included in the
+ * `hookNames` object. This is a Principle of Least Authority (POLA) measure
+ * to ensure that each RPC method implementation only has access to the
+ * API "hooks" it needs to do its job.
+ *
+ * @param hooks - The hooks to select from.
+ * @param hookNames - The names of the hooks to select.
+ * @returns The selected hooks.
+ * @template Hooks - The hooks to select from.
+ * @template HookName - The names of all hooks to select.
+ */
 export function selectHooks<
-  Hooks extends Record<keyof HookNames, unknown>,
-  HookNames extends Record<string, boolean>,
->(hooks: Hooks, hookNames: HookNames): SelectedHooks<Hooks, HookNames>;
-
-export function selectHooks<
-  T extends Record<keyof U, unknown>,
-  U extends Record<string, boolean>,
->(hooks: T, hookNames?: U) {
+  Hooks extends Record<string, unknown>,
+  HookName extends keyof Hooks,
+>(
+  hooks: Hooks,
+  hookNames?: Record<HookName, boolean>,
+): Pick<Hooks, HookName> | undefined {
   if (hookNames) {
     return Object.keys(hookNames).reduce((hookSubset, _hookName) => {
-      const hookName = _hookName as keyof SelectedHooks<T, U>;
+      const hookName = _hookName as HookName;
       hookSubset[hookName] = hooks[hookName];
       return hookSubset;
-    }, {} as SelectedHooks<T, U>);
```

Comment:
At this point the object is a _partial_ `Pick<Hooks, HookName>` but at the end it is `Pick<Hooks, HookName>`. So I had to cast both for the types to be correct _during_ the reduce function and after.